packages/core/src/command/consolidated/consolidated-create.ts

```diff
@@ -22,78 +22,104 @@ const emptyBundle: Bundle = {
 };
 
-export class PatientDataConsolidator {
-  readonly s3Utils: S3Utils;
-  constructor(private readonly consolidatedBucket: string, region: string) {
-    this.s3Utils = new S3Utils(region);
-  }
+export type ConsolidatePatientDataCommand = {
+  cxId: string;
+  patientId: string;
+  destinationBucketName?: string;
+  sourceBucketName?: string;
+  logMemUsage?: () => void;
+};
 
-  async execute({
-    cxId,
-    patientId,
-    inputBundles,
-    logMemUsage,
-  }: ConsolidatePatientDataCommand): Promise<Bundle> {
-    const { log } = out(`data consolidator - pat ${patientId}`);
+type BundleLocation = { bucket: string; key: string };
 
-    const getConsolidatedPromise = async (): Promise<NonNullableFields<Consolidated>> => {
-      log(`Getting existing consolidated data from bucket ${this.consolidatedBucket}`);
-      const consolidated = await getConsolidated({
-        cxId,
-        patientId,
-        fileLocation: this.consolidatedBucket,
-        region: this.s3Utils.region,
-        log,
-      });
-      log(`...consolidated key ${consolidated.fileName}`);
-      if (!consolidated.bundle) {
-        log(`No consolidated data found for patient ${patientId}, creating a new one`);
-      }
-      const returnBundle = consolidated.bundle ?? emptyBundle;
-      return { ...consolidated, bundle: returnBundle };
-    };
+/**
+ * Create a consolidated bundle from the existing conversion bundles.
+ */
+export async function createConsolidatedFromConversions({
+  cxId,
+  patientId,
+  destinationBucketName = getConsolidatedLocation(),
+  sourceBucketName = getConsolidatedSourceLocation(),
+  logMemUsage,
+}: ConsolidatePatientDataCommand): Promise<Bundle | undefined> {
+  const { log } = out(`createConsolidatedFromConversions - cx ${cxId}, pat ${patientId}`);
 
-    const {
-      bundle: consolidated,
-      fileLocation: destFileLocation,
-      fileName: destFileName,
-    } = await getConsolidatedPromise();
-    logMemUsage && logMemUsage();
+  const inputBundles = await listConversionBundles({ cxId, patientId, sourceBucketName, log });
+  if (!inputBundles || inputBundles.length < 1) {
+    log(`No conversion bundles found.`);
+    return undefined;
+  }
+
+  const consolidated = emptyBundle;
 
-    for (const inputBundle of inputBundles) {
+  await executeAsynchronously(
+    inputBundles,
+    async inputBundle => {
       const { bucket, key } = inputBundle;
-      log(`Getting input bundle from bucket ${bucket}, key ${key}`);
+      log(`Getting conversion bundle from bucket ${bucket}, key ${key}`);
       const contents = await executeWithRetriesS3(
-        () => this.s3Utils.getFileContentsAsString(bucket, key),
+        () => s3Utils.getFileContentsAsString(bucket, key),
         { ...defaultS3RetriesConfig, log }
       );
+      log(`Merging bundle ${key} into the consolidated one`);
       const bundle = JSON.parse(contents) as Bundle;
-      this.merge(bundle).into(consolidated);
+      merge(bundle).into(consolidated);
```

Comment:
is there any chance 10 things being merged into this bundle simaltaneously could cause issues? do we need a lock?

Reply:
t will be done in sequence, Node is single threaded… only the IO-based operations run in parallel